iotdb-core/datanode/src/main/java/org/apache/iotdb/db/subscription/broker/SubscriptionPrefetchingQueue.java

```diff
@@ -0,0 +1,253 @@
+        LOGGER.warn("Subscription: Ignore EnrichedEvent {} when prefetching.", event);
+      }
+    }
+
+    if (!tablets.isEmpty()) {
+      String subscriptionCommitId = generateSubscriptionCommitId();
+      SerializedEnrichedEvent enrichedEvent =
+          new SerializedEnrichedEvent(
+              new EnrichedTablets(topicName, tablets, subscriptionCommitId), enrichedEvents);
+
+      prefetchingQueue.add(enrichedEvent);
+
+      uncommittedEventsLock.writeLock().lock();
+      uncommittedEvents.put(subscriptionCommitId, enrichedEvent);
+      uncommittedEventsLock.writeLock().unlock();
+    }
+  }
+
+  private void serializeOnce() {
+    SerializedEnrichedEvent event;
+    try (ConcurrentIterableLinkedQueue<SerializedEnrichedEvent>.DynamicIterator iter =
+        prefetchingQueue.iterateFromEarliest()) {
+      // TODO: memory control
+      while (Objects.nonNull(
+          event =
+              iter.next(
+                  SubscriptionConfig.getInstance().getSubscriptionSerializeMaxBlockingTimeMs()))) {
+        event.serialize();
+      }
+    }
+  }
+
+  /////////////////////////////// recycle ///////////////////////////////
+
+  public void recycleUncommittedEvents() {
+    uncommittedEventsLock.writeLock().lock();
+    try {
+      recycleUncommittedEventsInternal();
+    } finally {
+      uncommittedEventsLock.writeLock().unlock();
+    }
+  }
+
+  private void recycleUncommittedEventsInternal() {
+    int count = 0;
+    while (true) {
+      if (count >= SubscriptionConfig.getInstance().getSubscriptionMaxEventsPerRecycling()) {
+        break;
+      }
+      if (uncommittedEvents.isEmpty()) {
+        break;
+      }
+      String subscriptionCommitId = uncommittedEvents.firstKey();
+      SerializedEnrichedEvent event = uncommittedEvents.get(subscriptionCommitId);
+      if (Objects.nonNull(event) && event.maybeExpired()) {
+        LOGGER.info(
+            "Subscription: recycle the event with subscriptionCommitId {}", subscriptionCommitId);
+        prefetchingQueue.add(event);
+        uncommittedEvents.remove(subscriptionCommitId);
+        count++;
+      }
+    }
+  }
+
+  /////////////////////////////// utility ///////////////////////////////
+
+  private Tablet convertToTablet(TabletInsertionEvent tabletInsertionEvent) {
+    if (tabletInsertionEvent instanceof PipeInsertNodeTabletInsertionEvent) {
+      return ((PipeInsertNodeTabletInsertionEvent) tabletInsertionEvent).convertToTablet();
+    } else if (tabletInsertionEvent instanceof PipeRawTabletInsertionEvent) {
+      return ((PipeRawTabletInsertionEvent) tabletInsertionEvent).convertToTablet();
+    }
+
+    LOGGER.warn(
+        "Subscription: Only support convert PipeInsertNodeTabletInsertionEvent or PipeRawTabletInsertionEvent to tablet. Ignore {}.",
+        tabletInsertionEvent);
+    return null;
+  }
+
+  private String generateSubscriptionCommitId() {
+    return topicName + "_" + brokerId + "#" + subscriptionCommitIdGenerator.getAndIncrement();
```

Comment:
```suggestion
    return topicName + "_" + brokerId + "#" + subscriptionCommitIdGenerator.getAndIncrement();
```

datanode id + reboot times?

Reply:
client should make sure that sending commit requests to where the data comes